src/ucp/core/ucp_worker.c

```diff
@@ -484,6 +484,27 @@ static ucp_ep_h ucp_worker_find_lane(ucs_list_link_t *ep_list, uct_ep_h uct_ep,
     return NULL;
 }
 
+/**
+ * FLUSH_CANCEL operation might be on pending queue due to
+ * UCS_ERR_NO_RESOURCES, so need to purge the queue to resubmit the
+ * operation. We need to resubmit the FLUSH_CANCEL operation on the same
+ * failed lane, in order to make sure all previous outstanding
+ * operations are completed before destroying the failed endpoint.
+ */
+static void ucp_worker_discard_lane_ff(ucp_worker_h worker, uct_ep_h uct_ep)
+{
+    ucs_queue_head_t tmp_q;
+    uct_pending_req_t *uct_req;
+
+    /* @ref ucs_arbiter_t does not support recursive calls, so use temporary
+     * queue */
+    ucs_queue_head_init(&tmp_q);
+    uct_ep_pending_purge(uct_ep, ucp_request_purge_enqueue_cb, &tmp_q);
+    ucs_queue_for_each_extract(uct_req, &tmp_q, priv, 1) {
+        ucp_ep_err_pending_purge(uct_req, UCS_STATUS_PTR(UCS_ERR_CANCELED));
```

Comment:
maybe call ucp_request_send_state_ff directly?

Reply:
it will add some code duplication also we are purging requests here, so IMO it makes sense to reuse existing function for this

Reply:
it's weird that this function is called "_ff" then it calls "err_pending_purge" which also calls "request_ff"
i think it just replace one call by another and make more sense with the naming. also, ucp_ep_err_pending_purge is supposed to be only a callback for uct_ep_pending_purge, and not be called directly.